### PR TITLE
Added implementation for LTE frequency bands get/set, network operator mode get/set

### PR DESCRIPTION
**Pull request by mdvermeer**
- removed LTE frequency band config from modem startup (UE init function) and instead allow combination of modem self-config and SIM specific config to determine frequency bands to use
- also added LwM2M disable on modem startup and removed frequency bands config

### CODE REVIEW

File: source/cellular_bg770_api.c [hunk 1]
```diff
@@ -450,20 +467,22 @@ static bool _parseSignalQuality( char * pQcsqPayload,
...
-                if( cellularStatus != CELLULAR_SUCCESS )
+                    if( cellularStatus != CELLULAR_SUCCESS )
+                    {
+                        atCoreStatus = CELLULAR_AT_BAD_PARAMETER;
+                    }
+                }
+                else
                 {
-                    atCoreStatus = CELLULAR_AT_BAD_PARAMETER;
+                    atCoreStatus = CELLULAR_AT_ERROR;
```
Comment (benwillard) [inline source/cellular_bg770_api.c:483]: This code seems to be unrelated to the purpose of the PR -- did you mean to include it?

Comment (benwillard) [reply source/cellular_bg770_api.c:483]: (Not a problem, just want to make sure it wasn't an oversight)

File: source/cellular_bg770_api.c [hunk 2]
```diff
@@ -6475,3 +6500,787 @@ CellularError_t Cellular_SetServiceSelection( CellularHandle_t cellularHandle,
...
+    bool isUpperNibble = false;
+    for( int i = bandHexStringLen - 1; i >= 0; i-- )
+    {
+        char tempBuffer[ 2 ] = { pTempBandHexString[ i ], '\0' };
+        char *pEnd = tempBuffer;
+        long tempVal = strtol( tempBuffer, &pEnd, 16 );
+        if( ( tempVal < 0x0 ) || ( tempVal > 0xF ) )
+        {
+            LogError( ( "_convertLTEFrequencyBandHexStringToBandBitmask non-hex character encountered" ) );
+            return false;
```
Comment (benwillard) [inline source/cellular_bg770_api.c:6648]: Can this ever happen? I thought `strtol` returned 0 if the value couldn't be parsed.

File: source/cellular_bg770_api.c [hunk 3]
```diff
@@ -6475,3 +6500,787 @@ CellularError_t Cellular_SetServiceSelection( CellularHandle_t cellularHandle,
...
+    memset( pLTEBandMask->asBytes, 0x00, LTE_BAND_MASK_LENGTH );
+
+    // loop through nibbles and apply to band bitmask
+    unsigned int byteIndex = LTE_BAND_MASK_LENGTH - 1;
+    bool isUpperNibble = false;
+    for( int i = bandHexStringLen - 1; i >= 0; i-- )
+    {
+        char tempBuffer[ 2 ] = { pTempBandHexString[ i ], '\0' };
+        char *pEnd = tempBuffer;
+        long tempVal = strtol( tempBuffer, &pEnd, 16 );
```
Comment (benwillard) [inline source/cellular_bg770_api.c:6644]: Why are we passing in the tempBuffer as both parameters here (aliased)?  `pEnd` is not useful for the parsing in the case of a single digit -- shouldn't we pass `NULL`?

Comment (benwillard) [reply source/cellular_bg770_api.c:6644]: I am also a bit worried about the aliasing since I believe both of the first two parameters to `strtol()` are marked as `restrict` parameters in the header.

Comment (benwillard) [reply source/cellular_bg770_api.c:6644]: (I *think* the aliasing is OK because one is a `const char* restrict` and the other is a `const char** restrict` but it's always hard to reason about)

File: source/cellular_bg770_api.c [hunk 4]
```diff
@@ -6475,3 +6500,787 @@ CellularError_t Cellular_SetServiceSelection( CellularHandle_t cellularHandle,
...
+        }
+
+        isUpperNibble = !isUpperNibble;
+    }
+
+    return true;
+}
+
+/* FreeRTOS Cellular Library types. */
+/* coverity[misra_c_2012_rule_8_13_violation] */
```
Comment (benwillard) [inline source/cellular_bg770_api.c:6686]: Are you running coverity on this code?

File: source/cellular_bg770_api.c [hunk 5]
```diff
@@ -6475,3 +6500,787 @@ CellularError_t Cellular_SetServiceSelection( CellularHandle_t cellularHandle,
...
+            // only record upper nibble if non-zero or previous non-zero value
+            if( noneZeroValueFound || upperNibble != 0 )
+            {
+                pBandHexString[ charIndex ] = _getHexCharFromNibble( upperNibble );
+                charIndex++;
+            }
+
+            pBandHexString[ charIndex ] = _getHexCharFromNibble( pLTEBandMask->asBytes[ i ] & 0xF );
+            charIndex++;
+
```
Comment (benwillard) [inline source/cellular_bg770_api.c:6831]: Seems like a lot of work that `snprintf()` could do for you

File: source/cellular_bg770_api.c [hunk 6]
```diff
@@ -6475,3 +6500,787 @@ CellularError_t Cellular_SetServiceSelection( CellularHandle_t cellularHandle,
...
+        if( origDesiredFrequencyBandsByte != pDesiredFrequencyBands->asBytes[ i ] )
+        {
+            bandFiltered = true;
+        }
+    }
+
+    if( out_bandFiltered != NULL )
+    {
+        *out_bandFiltered = bandFiltered;
```
Comment (benwillard) [inline source/cellular_bg770_api.c]: Confusing naming, `*out_bandChanged` would be my suggestion

File: source/cellular_bg770_api.c [hunk 7]
```diff
@@ -6475,3 +6500,787 @@ CellularError_t Cellular_SetServiceSelection( CellularHandle_t cellularHandle,
...
+    for( int i = 0; i < LTE_BAND_MASK_LENGTH; i++ )
+    {
+        if( pFrequencyBands->asBytes[ i ] != 0 )
+        {
+            return true;
+        }
+    }
+
+    return false;
+}
```
Comment (benwillard) [inline source/cellular_bg770_api.c]: Why doesn't this live in the module that defines `CellularLTEBandMask_t`? This does not seem bg770 specific.

File: source/cellular_bg770_api.c [hunk 8]
```diff
@@ -6475,3 +6500,787 @@ CellularError_t Cellular_SetServiceSelection( CellularHandle_t cellularHandle,
...
+            {
+                LogError( ( "Cellular_SetLTEFrequencyBands : Specified frequency bands not supported by BG770" ) );
+            }
+            else
+            {
+                if( bandsFiltered )
+                {
+                    LogWarn( ( "Cellular_SetLTEFrequencyBands : Unsupported LTE frequency bands removed" ) );
+                }
+            }
```
Comment (benwillard) [inline source/cellular_bg770_api.c]: `else if` or is this a MISRA thing?

File: source/cellular_bg770_api.c [hunk 9]
```diff
@@ -6475,3 +6500,787 @@ CellularError_t Cellular_SetServiceSelection( CellularHandle_t cellularHandle,
...
+        return false;
+    }
+
+    for( int i = 0; cstring[ i ] != '\0'; i++ )
+    {
+        cstring[ i ] = tolower( ( int ) cstring[ i ] );
+    }
+
+    return true;
+}
```
Comment (benwillard) [inline source/cellular_bg770_api.c:7025]: Why not `string.h`'s `strlwr()`?

**Review by benwillard — APPROVED** [review]
Listed some minor things to go through. Approving since I won't be around tomorrow.
One thing that surprised me was no unit tests -- in my experience anything doing string parsing usually benefits very quickly from unit tests.

File: source/cellular_bg770_api.c (hunk 1 above)
Comment (mdvermeer) [reply source/cellular_bg770_api.c:483]: Yes, it's not directly related to the PR's main purpose but was discovered while making changes. Not breaking it out into it's own PR was a little lazy.

File: source/cellular_bg770_api.c (hunk 3 above)
Comment (mdvermeer) [reply source/cellular_bg770_api.c:6644]: pEnd is useful since it allows differentiation between 0 because of error and real zero. pEnd == NULL or *pEnd not null terminator indicates error conditions since we cannot look at unsafe errno.

File: source/cellular_bg770_api.c (hunk 4 above)
Comment (mdvermeer) [reply source/cellular_bg770_api.c:6686]: I haven't been. But I've been moving the comments because they come from code doing nearly the same thing.

File: source/cellular_bg770_api.c (hunk 2 above)
Comment (mdvermeer) [reply source/cellular_bg770_api.c:6648]: According to the documentation it shouldn't but there's also mention of undefined behaviour under certain circumstances. I'm covering all bases. This condition just ensures if we return a true for the function then the bitmask is rock solid.

File: source/cellular_bg770_api.c (hunk 5 above)
Comment (mdvermeer) [reply source/cellular_bg770_api.c:6831]: The overhead from snprintf for this is large. Plus I don't want any leading zeroes.

File: source/cellular_bg770_api.c (hunk 7 above)
Comment (mdvermeer) [reply source/cellular_bg770_api.c]: I can move it. Generally the FreeRTOS-Cellular-Interface doesn't include these type of analysis functions in the API.

Comment (mdvermeer) [reply source/cellular_bg770_api.c]: CellularLTEBandMask_t is defined in a types .h (just following the existing API structure). So there's no .h/.c pair for me to put this in.

File: source/cellular_bg770_api.c (hunk 8 above)
Comment (mdvermeer) [reply source/cellular_bg770_api.c]: Not a MISRA thing, just changed structure and forgot to consolidate. Will change

File: source/cellular_bg770_api.c (hunk 6 above)
Comment (mdvermeer) [reply source/cellular_bg770_api.c]: Changed to out_bandmaskChanged

File: source/cellular_bg770_api.c (hunk 9 above)
Comment (mdvermeer) [reply source/cellular_bg770_api.c:7025]: strlwr() is non-ANSI standard and therefore I stuck with code that is standard.
